src/components/timeline/LayerDiscrete.svelte

```diff
@@ -125,34 +126,52 @@
   $: onMousedown(mousedown);
   $: onMousemove(mousemove);
   $: onMouseout(mouseout);
-  $: onMouseup(mouseup);
+  $: onMouseUp(mouseup);
 
   $: anchorIconWidth = 16;
   $: anchorIconMarginLeft = 4;
   $: canvasHeightDpr = drawHeight * dpr;
   $: canvasWidthDpr = drawWidth * dpr;
-  $: rowHeight = activityOptions.activityHeight + (activityOptions.displayMode === 'compact' ? 0 : 0);
+  $: rowHeight = discreteOptions.height + (discreteOptions.displayMode === 'compact' ? 0 : 0);
   $: timelineLocked = timelineLockStatus === TimelineLockStatus.Locked;
   $: planStartTimeMs = getUnixEpochTime(getDoyTime(new Date(planStartTimeYmd)));
 
-  $: if (
-    activityDirectives &&
-    showDirectives !== undefined &&
-    showSpans !== undefined &&
-    canvasHeightDpr &&
+  // the following are NOT mutually exclusive.
+  $: commonConditions = canvasHeightDpr &&
     canvasWidthDpr &&
     ctx &&
     drawHeight &&
     drawWidth &&
     dpr &&
+    discreteOptions &&
+    viewTimeRange &&
+    xScaleView &&
+    discreteTree
+
+  $: canDrawActivities = activityDirectives &&
+    showDirectives !== undefined &&
+    showSpans !== undefined &&
     selectedActivityDirectiveId !== undefined &&
     selectedSpanId !== undefined &&
     spansMap &&
-    activityOptions &&
-    viewTimeRange &&
-    xScaleView &&
-    spans &&
-    activityTree
+    discreteOptions.activityOptions &&
+    spans;
+
+  $: canDrawExternalEvents = externalEvents &&
+    selectedExternalEventId !== undefined &&
+    discreteOptions.externalEventOptions
+
+  $: if (
+    commonConditions && (canDrawActivities || canDrawExternalEvents) && xScaleView
+  ) {
+    // if this print is excluded, contents of row vanish on row height change for some reason and draw is not executed. Including it is necessary?
+    console.debug("Row Height updated: ", drawHeight); 
+    draw();
+  }
```

Comment:
Huh, that's wacky. If I were to venture a guess, it's the fact that you're referencing `drawHeight` in this line of code that makes all the difference. Svelte reactive statements are rerun whenever any of the variables that they reference are changed. Importantly, these variables need to be syntactically part of the `$:` statement - the compiler doesn't inspect the `draw` function to figure out what it depends on.

Some docs here: https://svelte.dev/docs/svelte-components#script-3-$-marks-a-statement-as-reactive

Try tacking it on to the end of the condition like this and see if that works (note the `&& drawHeight` at the end there):

```suggestion
  $: if (
    commonConditions && (canDrawActivities || canDrawExternalEvents) && xScaleView && drawHeight
  ) {
    draw();
  }
```

Reply:
I'm going to try your recommendation out but is there any reason putting it in the `if` condition as opposed to where it is now (included in that `commonConditions` boolean) would change things? Does including it in a boolean just mean it only investigates changes when drawHeight changes from 0 to non-zero as opposed to non-zero to non-zero?

Reply:
In any case, it seems like your suggestion resolved it; see (this commit)[https://github.com/NASA-AMMOS/aerie-ui/commit/f279e8a82f3d7c47e8caead9d4490e13aba3e447].

Reply:
@pranav-super yes, I think you're right. A svelte reactive statement only re-runs if one of the variables it references _changes_.

In a bit more detail, my mental model is that every time svelte runs the reactive statement, it records which reactive variables were actually read on that run (which will be a subset of those that are syntactically part of the statement). It will only rerun if the value of one of those variables changes.

Changing `drawHeight` causes `commonConditions` to be recomputed - but if it was `true` before, and it's `true` now, then its value hasn't changed, so it will not trigger anything else to rerun.

Unfortunately I think that makes the `commonConditions` pattern a bit suspect - I don't think it's doing what we want it to do (at least, I'm assuming we want it to re-trigger other reactive statements whenever any of its operands change - correct me if I'm wrong).

I'll see if there are any recommendations for best practices here... the only idea I have off the top of my head is hashing the set of inputs so that the value of commonConditions actually changes 🤔

Reply:
I'll summarize the situation below for the benefit of anyone following along (and to help me get my thoughts straight).

I think it comes down to enumerating the two different use cases we have here:
- (1) The condition part of this reactive statement protects the draw function from being called when one of its dependencies hasn't been initialized yet. This is why we make sure that all of the reactive variable values are "truthy"
- (2) After everything has been initialized, we want to re-trigger the draw function whenever any of the dependencies change

We get both properties if we only have one reactive statement:
```js
// Listing 1
$: if ((canvasHeightDpr &&
       canvasWidthDpr &&
       ctx &&
       drawHeight &&
       drawWidth &&
       dpr &&
       discreteOptions &&
       viewTimeRange &&
       xScaleView &&
       discreteTree) && (canDrawActivities || canDrawExternalEvents)) {
  draw();
}
```

To clean up the code, we've extracted that first clause into a new reactive variable called `commonConditions`:
```js
// Listing 2
$: commonConditions = canvasHeightDpr &&
       canvasWidthDpr &&
       ctx &&
       drawHeight &&
       drawWidth &&
       dpr &&
       discreteOptions &&
       viewTimeRange &&
       xScaleView &&
       discreteTree

$: if (commonConditions && (canDrawActivities || canDrawExternalEvents)) {
  draw();
}
```
This implementation preserves property (1) above, but it fails property (2): after all the variables have become "truthy", any subsequent modifications of a variable from a truthy value to another truthy value will NOT trigger the `draw` function.

We work around this by selectively adding back the properties that we expect to change - so far that's `xScaleView` and `drawHeight`, but of course it begs the question of whether we've missed anything:
```js
// Listing 3
$: commonConditions = canvasHeightDpr &&
       canvasWidthDpr &&
       ctx &&
       drawHeight &&
       drawWidth &&
       dpr &&
       discreteOptions &&
       viewTimeRange &&
       xScaleView &&
       discreteTree

$: if (commonConditions && (canDrawActivities || canDrawExternalEvents) && xScaleView && drawHeight) {
  draw();
}
```

I'll pause here and refrain from offering solutions until @duranb has had a chance to weigh in.

Reply:
I would say that boiling it all down into a `commonConditions` might be oversimplifying it. As @mattdailis already mentioned, this only captures the `isDefined` condition, but not the `hasChanged` part of this.

If you would like to capture those explicitly, then I would suggest:
1. Naming the variable more explicitly (e.g. `areAssetsDefined` instead of `commonConditions`)
2. Making sure that only the valid variables that determine the final state of `commonConditions` are used so as not to infer the wrong update logic. The way it is currently used here implies that each variable can go from a falsy to truthy state and might mislead the dev reading this to think that the variables are just initialized once or something.

I think it's actually more clear to just keep it all listed in the `if` conditional. It's a common svelte-ism imo. If you wanted to have a single variable track all the changes, you would have to then track the previous states of all the variables so that you have something to compare to, which is probably not worth it.

Reply:
The goal was mainly to clean up the code and make it more readable. Expanding to what Matt suggests in Listings 1 or 3 is logically equivalent to what we were trying to express, except those listings actually work. As for which we should go with, I believe it is a matter of preference; is there a particular one you would prefer?

Reply:
> I think it's actually more clear to just keep it all listed in the if conditional. It's a common svelte-ism imo

@pranav-super my understanding of Bryan's comment is he's recommending Listing 1, which includes all of the relevant variables directly in the conditional. Listing 3 is still suspect - e.g. what happens when drawWidth changes?

Let's go ahead and inline `commonConditions` and remove the variable, and see how the result looks and assess readability.

Reply:
Alright! I updated it [here](https://github.com/NASA-AMMOS/aerie-ui/pull/1396/commits/ce9bb8f7e0a73ab99e8c829ac1f25d8075131799). Let me know if this looks good, I think it seems alright.

Reply:
Looks good to me, thanks for addressing that :+1: Let's keep an eye out for similar tricky uses of reactivity